src/functions_math.jl

```diff
@@ -78,7 +78,20 @@ macro declare_matmul(MatrixT, VectorT=nothing)
     return esc(Expr(:block, codes...))
 end
 
+# The following two methods can be defined by using
+# @declare_matmul(Diagonal, AbstractVector)
+# but that overwrites existing *(1D NDA, Vector) methods
+# should improve the macro above to deal with this case
+function Base.:*(a::Diagonal, b::NamedDimsArray{<:Any, <:Any, 1})
+    return *(NamedDims.NamedDimsArray{dimnames(a)}(a), b)
+end
```

Comment:
wrapping a Diagonal in a NDA does seem a roundabout way to do this... i don't have a better proposal, but it makes me suspect something is still not quite right somewhere

Reply:
This is what the `declare_matmul` macro does as well. I think the idea is to reuse the NDA * NDA method, which deals with the naming of output dimensions. I guess you are worried about the conversion overhead?

Reply:
No, there shouldn't be an overhead - I was his surprised by the approach